### ISSUE TITLE
ngrx 4.0 and StoreModule.forFeature

### ISSUE THREAD

**Issue by etatuev** [btroncone/ngrx-store-localstorage#44]
Need a way to correctly serialize separately loaded slices of state (I mean from `StoreModule.forFeature(...)` in lazy loaded module). Problem is that currently we must to set keys which will be serialized at the metaReducer creation, but we can't know lazy loaded keys. Liar, we can make a hack and hardcode all keys at the start. But what if we have dynamic modules? (my case)
 
Other way is to apply metaReducer for each `forFeature` state, but then we cannot access root state, but rather see only 2nd level objects.

Example (pseudo):
```
class RootState {
  core: ...
  auth: ...
  // here will be lazy loaded state, not known at this stage
  // lazy1: ...
}
```

`StoreModule.forRoot (RootState, meta: sync(keys: ['core', 'auth']) ) // root state object available at metaReducer, everything is OK `

```
class LazyState {
  page1: ...
  page2: ...
}
```

`StoreModule.forFeature ('lazy', LazyState, meta: sync(keys ['lazy']) ) //we want to set only top level keys to avoid storage key conflicts, and reduce count of serialized objects`

But what happens -  `rootState.lazy` object available at `syncStateUpdate \ rehydrate` instead of `rootState`.  So, metaReducer tried to iterate over keys (`['lazy']` in that case), but available state object has only `page1` and `page2` props. Other words, metaReducers for `forFeature` has access only for current 'child' state.

Possible workaround is to make "keys" options of localstorage-sync available "globally", to be able to extend it at runtime. Currently `validateStateKeys` break that workaround because of `map`, which create new array, and we cannot access it by initial reference. I can prepare PR which change `map` to `forEach` (not really matters for that func, but we can keep the array ref). But it sounds **VERY** ugly. Do you have some thoughts, guys?

**UPD:**
Currently my workaround is to store child states for lazy loaded modules. 
And use my previously submitted `customKeySerializer` to avoid naming conflicts between child states :)
I set function like 
```
const prefix = 'somePrefix';
const func = (key) =>'${prefix}_${key}'
``` 
and set prefix as **root** for root slices, and **featureName** for lazy loaded states, together with passing child states as keys. 
For previous example (pseudo, again):
```
StoreModule.forRoot (RootState, meta: sync(keys: ['core', 'auth'], prefix:'root'))
StoreModule.forFeature ('lazy1', LazyState, meta: sync(keys: ['page1', 'page2'], prefix: 'lazy1') )
StoreModule.forFeature ('lazy2', LazyState, meta: sync(keys: ['page1', 'page2'], prefix: 'lazy2') )
```
In the end I have next localStorage keys:
```
root_core
root_auth
lazy1_page1  // should be just inside 'lazy1' prop
lazy1_page2  // should be just inside 'lazy1' prop
lazy2_page1  // should be just inside 'lazy2' prop
lazy2_page2  // should be just inside 'lazy2' prop
```
It works, but problem is if lazy loaded keys has a lot of child keys, we need to run serialize func and access Storage for many times, and it's a much slower than write single object at once.

**Comment by x3igh7** [comment]
Running into the same issue as you. I have a few modules with their own stores being lazy loaded and unable to get them to sync using the localStorage module. Might also poke around and see if there is a solution I can come up with.

**Comment by etatuev** [comment]
Yeah, it will be good, because I still use ugly workaround I mentioned, with applying this metareducer for every lazy loaded state separately :)

**Comment by kuncevic** [comment]
Just facing same issue and trying to figure out the best options.
Any updates on that?

**Comment by tanyagray** [comment]
The PR for this has now been merged and released as `v5.0.0`

Rehydration for feature modules is now supported, thanks @nhaesler !
